### PR TITLE
[CARBONDATA-1015] Refactory write step and add ColumnPage in data load

### PR DESCRIPTION
**Pull request by jackylk**
1. Add ColumnPage and TablePage to represent columnar data
2. Simplify write step in data loading (`CarbonFactDataHandlerColumnar.processDataRows`)
3. Unified data type enum, avoid using char as datatype

Main change is in `CarbonFactDataHandlerColumnar.processDataRows`, please start reviewing from this function.

No test case is added. Just use existing test case to verify the change.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1800/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1857/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1860/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1864/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1871/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1872/

File: core/src/main/java/org/apache/carbondata/core/datastore/page/FixLengthColumnPage.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+public class FixLengthColumnPage extends ColumnPage {
+
+  // Only one of following fields will be used
+  private byte[] byteData;
+  private short[] shortData;
+  private int[] intData;
+  private long[] longData;
+  private double[] doubleData;
+
+  private byte[][] byteArrayData;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/FixLengthColumnPage.java]: why fix length page has two dimensional array?

File: core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/PageStatistics.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+import org.apache.carbondata.core.metadata.datatype.DataType;
+import org.apache.carbondata.core.util.DataTypeUtil;
+
+/** statics for one column page */
+public class PageStatistics {
+  private DataType dataType;
+
+  /** min and max value of the measures */
+  private Object min, max;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/PageStatistics.java]: what about min/max for no dictionary column and dictionary column?

File: core/src/test/java/org/apache/carbondata/core/util/CarbonMetadataUtilTest.java [hunk 3]
```diff
@@ -169,7 +171,12 @@
     long[] longArr = { 1, 2, 3, 4, 5 };
     byte[][] maxByteArr = { { 1, 2 }, { 3, 4 }, { 5, 6 }, { 2, 4 }, { 1, 2 } };
     int[] cardinality = { 1, 2, 3, 4, 5 };
-    char[] charArr = { 'a', 's', 'd', 'g', 'h' };
+    org.apache.carbondata.core.metadata.datatype.DataType[] charArr = {
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/util/CarbonMetadataUtilTest.java]: better rename to ```dataTypes```

**Comment by ravipesala** [comment]
@jackylk Still wiring is not done for columnpage to writer step right? are you planning to do in this PR only?

File: core/src/main/java/org/apache/carbondata/core/datastore/page/FixLengthColumnPage.java (hunk 1 above)
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/FixLengthColumnPage.java]: It is for decimal data type. I will add comment for it.
I could not think of a better name for this class, could you suggest one?

File: core/src/test/java/org/apache/carbondata/core/util/CarbonMetadataUtilTest.java (hunk 3 above)
Comment (jackylk) [inline core/src/test/java/org/apache/carbondata/core/util/CarbonMetadataUtilTest.java]: ok

File: core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/PageStatistics.java (hunk 2 above)
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/PageStatistics.java]: The min and max for no dictionary and dictionary columns actually are get from `BlockIndexerStorageForInt` and `BlockIndexerStorageForShort` after each column is sorted. So I am not updating them here. Make sense?

**Comment by jackylk** [comment]
I have not changed Writer interface, so writer is not using ColumnPage yet.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1891/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1893/

**Pull request closed without merging**